Replace swagger-node-express with openapi-backend

anttiviljami opened the pull request:
This PR replaces the abandoned `swagger-node-express` module for [`openapi-backend`](https://www.npmjs.com/package/openapi-backend).

With this change, the API now exposes the openapi definition under `/api/openapi.json` (for the latest api version). Previous api versions are available for each api root `/api/{version}/openapi.json`.

These changes should bring no other functional changes to the existing HTTP API.

anttiviljami commented:
This implementation is also now referenced in the official `openapi-backend` docs. 🎉 https://github.com/anttiviljami/openapi-backend/blob/master/DOCS.md#api-versioning

anttiviljami commented:
Looks like travis is currently failing in `develop` branch. I can rebase if needed, once this is fixed.

JohnMcLear commented:
Travis is fickle.  Afaik one of the repo owners have to send through the PR in order for tests to run.

I will play with this now :)

JohnMcLear commented:
- [x] backend tests pass 
``npm run test`` looks good
 - [ ] testing frontend (not sure why it matters).

JohnMcLear commented:
Okay over to @muxator for code review :)  I'm +1!  Tests pass.

1 request, update the docs to mention the new openapi.json endpoint et al?

JohnMcLear commented:
Oh ``http://192.168.1.51:9001/api/openapi.json`` is returning 

```
Cannot GET /api/openapi.json
``` 

?

anttiviljami commented:
> Cannot GET /api/openapi.json

Hmm, not sure what that could be. Works on my end. 

I found a small issue with a `$ref`. Maybe pull, restart and try again?

<img width="524" alt="Screenshot 2020-03-29 at 22 13 20" src="https://user-images.githubusercontent.com/6105650/77859575-841db400-720a-11ea-8a45-14fbf127eca8.png">

anttiviljami commented:
> 1 request, update the docs to mention the new openapi.json endpoint et al?

Already added this. Any other places this should be mentioned? https://github.com/ether/etherpad-lite/pull/3786/files#diff-04c6e90faac2675aa89e2176d2eec7d8R114

JohnMcLear commented:
I'm an idiot 🗡 
``npm run test``
``Error: Cannot find module 'openapi-schema-validation'``

JohnMcLear commented:
```
[2020-03-29 20:17:26.449] [ERROR] console - Failed to load '/home/jose/anttiviljamietherpad/node_modules/ep_etherpad-lite/node/hooks/express/openapi:expressCreateServer' for 'ep_etherpad-lite/openapi/hooks/expressCreateServer': Error: Cannot find module 'lodash/clonedeep'
Require stack:
- /home/jose/anttiviljamietherpad/src/node/hooks/express/openapi.js
- /home/jose/anttiviljamietherpad/src/static/js/pluginfw/shared.js
- /home/jose/anttiviljamietherpad/src/static/js/pluginfw/plugins.js
- /home/jose/anttiviljamietherpad/src/node/server.js
[2020-03-29 20:17:26.950] [INFO] APIHandler - Api key file read from: "/home/jose/anttiviljamietherpad/APIKEY.txt"
[2020-03-29 20:17:27.115] [INFO] console - Installed plugins:
[2020-03-29 20:17:27.117] [ERROR] console - Failed to load 'ep_etherpad-lite/node/hooks/express/openapi:expressCreateServer' for 'ep_etherpad-lite/openapi/hooks/expressCreateServer': Error: Cannot find module 'lodash/clonedeep'
Require stack:
- /home/jose/anttiviljamietherpad/src/node/hooks/express/openapi.js
- /home/jose/anttiviljamietherpad/src/static/js/pluginfw/shared.js
- /home/jose/anttiviljamietherpad/src/static/js/pluginfw/plugins.js
- /home/jose/anttiviljamietherpad/src/node/server.js

```

JohnMcLear commented:
Missing some deps from packages.json? :P

anttiviljami commented:
> Error: Cannot find module 'openapi-schema-validation'
> Error: Cannot find module 'lodash/clonedeep'

Hmm, these should all be bundled with openapi-backend. 🤨

anttiviljami commented:
I'll double check!

anttiviljami commented:
ff68642 should help!

JohnMcLear commented:
no dice ;\

JohnMcLear commented:
```
jose@server:~/anttiviljamietherpad/src$ npm run test

> ep_etherpad-lite@1.8.0 test /home/jose/anttiviljamietherpad/src
> nyc mocha --timeout 5000 ../tests/backend/specs/api


Error: Cannot find module 'openapi-schema-validation'
Require stack:
```

anttiviljami commented:
what the... must be something wrong with your environment. it's in the lock file just fine

https://github.com/ether/etherpad-lite/blob/ff68642bd05433dac4848a9c524c54cc697afc85/src/package-lock.json#L6362

anttiviljami commented:
```
rm -fr node_modules
npm install
```

?

JohnMcLear commented:
```
Error: Cannot find module 'openapi-schema-validation'
Require stack:
```

Checking my files..

anttiviljami commented:
wait, are you running tests with `npm install --production` or `NODE_ENV=production` ?

JohnMcLear commented:
``package.json``
```
  "devDependencies": {
    "mocha": "7.1.1",
    "nyc": "15.0.0",
    "openapi-schema-validation": "^0.4.2",
```

JohnMcLear commented:
in ``package-lock.json``

```
    "@apidevtools/openapi-schemas": {
      "version": "2.0.3",
      "resolved": "https://registry.npmjs.org/@apidevtools/openapi-schemas/-/openapi-schemas-2.0.3.tgz",
      "integrity": "sha512-QoPaxGXfgqgGpK1p21FJ400z56hV681a8DOcZt3J5z0WIHgFeaIZ4+6bX5ATqmOoCpRCsH4ITEwKaOyFMz7wOA=="
    },
```

JohnMcLear commented:
``npm run test``
not setting any environment variables.

JohnMcLear commented:
fwiw /openapi.json route working now :)

anttiviljami commented:
Fresh ubuntu install on node 10

![test](https://user-images.githubusercontent.com/6105650/77860346-060fdc00-720f-11ea-9f14-cf8f347c7910.gif)

JohnMcLear commented:
You didn't ``git checkout feature/openapi-backend`` 💃

JohnMcLear commented:
https://github.com/ether/etherpad-lite/blob/develop/doc/api/http_api.md I guess

anttiviljami commented:
Sorry! eb26578 should clear up the dependency issue. Apparently mocha doesn't share its path with rest of the project. I must've installed something openapi related globally so it worked on my machine.

Added a section about exposing the openapi definition in 4d04550 !

JohnMcLear commented:
Whoop nice, looks good and passes all tests.  Over to @muxator

muxator commented:
I'll have a look now.

muxator commented:
Quick feedback.

* correctly, this new implementation keeps untouched the test specifications in `tests/backend/specs/api`
* a new spec file [`tests/backend/specs/api/api.js`](https://github.com/anttiviljami/etherpad-lite/blob/feature/openapi-backend/tests/backend/specs/api/api.js) is introduced. @anttiviljami, could you please give a short description of its purpose, so we can agree on it and you can put it at the start of the spec file?
* **the backend tests pass just like before**. Good work!
* having moved to a maintained library, **we now no longer have any high risk security vulnerabilities**

I'll have a look at the changes now (and probably tomorrow).

@anttiviljami, are you comfortable if I force-push on your branch?
Are you finished, apart from the comment on `api.js` and any future modifications we might have to do?

muxator commented on src/node/hooks/express/openapi.js:
```diff
@@ -0,0 +1,666 @@
+
+      // serve version specific openapi definition
+      app.get(`${apiRoot}/openapi.json`, (req, res) => {
+        res.header('Access-Control-Allow-Origin', '*');
```
Potential big red flag here.

I am not really sure allowing CORS from everywhere would pass a security assessment.

Probably this needs to be configurable in the settings (and, jost out of my head, we have to test embedding works afterwards).

More work needed here.

anttiviljami replied:
For exposing the OAS definitions allowing CORS from everywhere is probably ok.

muxator commented on src/node/hooks/express/openapi.js:
```diff
@@ -0,0 +1,666 @@
+      const isLatestAPIVersion = version === apiHandler.latestApiVersion;
+      if (isLatestAPIVersion) {
+        app.get(`/${style}/openapi.json`, (req, res) => {
+          res.header('Access-Control-Allow-Origin', '*');
```
Same here.

`Access-Control-Allow-Origin: *` is convenient when developing, but wee cannot leave it hardcoded.

anttiviljami replied:
The reasoning for doing this here was purely to maintain existing behaviour and a great deal of effort was put into this PR to not introduce any breaking changes.

In fact it looks like all API paths always return `Access-Control-Allow-Origin: *` at the moment: https://github.com/ether/etherpad-lite/blob/2b753f13ea3dac02132033b3ab0be0dce3931c19/src/node/hooks/express/swagger.js#L435-L437 

I agree that this is not good practice but I would still propose that we fix this as part of a separate PR.

muxator replied:
Woah. It's really there.

I agree: it is better for this series to be a refactoring in the strictest sense. Bug-for-bug, vuln-for-vuln.

anttiviljami replied:
Looks like I was wrong. Only api paths starting with `/rest/{version}/**` seem to actually return `access-control-allow-origin: *` right now, even though `swagger-node-express` seems to clearly be configured to manage api paths. 

I'm very confused why this is but it looks like `/api/**` don't expose any cors headers in `develop`.

I'm unsure of what to do here. Should we just replicate this odd behaviour for this PR?

anttiviljami replied:
I've just replicated the behaviour from `develop` branch for now. Feel free to push to this branch to change cors behaviour!

Commit: 4fda128

anttiviljami commented:
> @anttiviljami, are you comfortable if I force-push on your branch?
> Are you finished, apart from the comment on api.js and any future modifications we might have to do?

Added the comment for the api spec file.

Feel free to force push to the branch. I'm finished on my end unless there's something you want me to still check.

As for the CORS issue, like I commented above, we can of course change the behaviour but that might introduce a breaking change for existing API users.

anttiviljami commented:
@muxator anything you still need from me for this to be mergeable?

muxator commented:
Next time, please do not mix functional changes and linting changes, as it makes really difficult to see what happens.

For example this: a37aa605fb8608009d31bea367e40c45b51d2692 change was probably (partially) reverted afterwards, with this b03f5509b569d096dc58802a106e3bc17b184609. Guessing these sort of things takes a lot of time.

Speaking of that changes, I guess APIHandler was totally reverted, while openapi.js was not, I think. Can you give some context?

muxator commented:
I went on and nulled the changes to `APIHandler` in a37aa605fb8608009d31bea367e40c45b51d2692 ("Add commented out version of response processors"). This simplifies c06490bb5867 ("bring back logging")

anttiviljami commented:
> Next time, please do not mix functional changes and linting changes, as it makes really difficult to see what happens.

I tried not to, but my editor ran prettier automatically at some point, which I didn’t immediately notice and later reverted the linting changes. 

> Speaking of that changes, I guess APIHandler was totally reverted, while openapi.js was not, I think. Can you give some context?

openapi.js is a completely new file so I let prettier do it’s thing there. 

Thanks for the cleanup!

muxator commented on src/node/hooks/express/openapi.js:
```diff
@@ -0,0 +1,671 @@
+          let response = { code: 0, message: 'ok', data };
+
+          // log response
+          apiLogger.info(`RESPONSE, ${funcName}, ${JSON.stringify(response)}`);
```
A doubt: what does happen when there is a failure? Is the API going to return a non zero code (and a relevant message)?
I am not sure the tests cover this case, and from the look of this part there could be this risk.

For your reference, the documentation for the HTTP API response format is in [doc/api/http_api.md](https://github.com/ether/etherpad-lite/blob/4ee5ddb496bdd3ab107efa26a001a87ac343fd76/doc/api/http_api.md#response-format), this is the relevant section:

> ### Response Format
> Responses are valid JSON in the following format:
> 
> ```json
> {
>   "code": number,
>   "message": string,
>   "data": obj
> }
> ```
> 
> * **code** a return code
>   * **0** everything ok
>   * **1** wrong parameters
>   * **2** internal error
>   * **3** no such function
>  * **4** no or wrong API Key
> * **message** a status message. It's ok if everything is fine, else it contains an error message
> * **data** the payload

Could you please confirm if we are respecting this?

anttiviljami replied:
Yes, we cover all the same cases as the previous implementation. 

1. If the api function throws an error, we either pass the custom error thrown by the function as code 1 or code 2 for non-`apierror` exceptions.

Previously: https://github.com/ether/etherpad-lite/blob/master/src/node/handler/APIHandler.js#L209-L219
Now: https://github.com/ether/etherpad-lite/blob/427d388631fceeacdec4dbc97933f35540655bf2/src/node/hooks/express/openapi.js#L650-L659

2. If the api definition doesn't contain the route being called, we return code 3 for no such function using the `notFound` handler:

https://github.com/ether/etherpad-lite/blob/427d388631fceeacdec4dbc97933f35540655bf2/src/node/hooks/express/openapi.js#L585-L588

3. For code 4, `Unauthorized`, there's a test which we pass

Test: https://github.com/ether/etherpad-lite/blob/master/tests/backend/specs/api/pad.js#L51-L59

muxator replied:
ok, thanks

anttiviljami commented:
> 1. I have a doubt about what happens in case of failures (see the comment in the code).

Addressed this above. This PR matches the current implementation. Feel free to add tests to `develop` so we can rebase this if further confirmation is required.

> 2. squashed some commits that where the result of progressive development (mistakes, bad dependencies).

Great! 👍

muxator commented:
The change on `tests/backend/specs/api/chat.js` here 57b38d14d7a9 is reverted in the next commit de87779fcf4b.

Was that a test?

anttiviljami commented:
> Was that a test?

No. The test was just moved to the new `test/backend/specs/api/api.js` file in https://github.com/ether/etherpad-lite/commit/de87779fcf4b2edb5f842e2836943926dd7f1511

muxator commented:
> No. The test was just moved to the new test/backend/specs/api/api.js file in de87779

Ok. Removed the modification & subsequent revert of `chat.js`, and kept in place the modification of `api.js`.

muxator commented:
A final remark, @anttiviljami: although all the tests pass, there is an error in the logs.
It is not easy to spot because the loags are overwelmed. 

```
[2020-04-01 13:43:41.601] [INFO] API - REQUEST, v1.2.13:createPad, {"apikey":"password","padID":"test"}
[2020-04-01 13:43:41.602] [INFO] API - RESPONSE, createPad, {"code":0,"message":"ok","data":null}
[2020-04-01 13:43:41.606] [ERROR] console - (node:18540) UnhandledPromiseRejectionWarning: Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:485:11)
    at ServerResponse.header (/mnt/md/tmp/etherpad-anttiviljami/src/node_modules/express/lib/response.js:771:10)
    at ServerResponse.send (/mnt/md/tmp/etherpad-anttiviljami/src/node_modules/express/lib/response.js:170:12)
    at ServerResponse.res.send (/mnt/md/tmp/etherpad-anttiviljami/src/node/hooks/express/webaccess.js:105:14)
    at ServerResponse.json (/mnt/md/tmp/etherpad-anttiviljami/src/node_modules/express/lib/response.js:267:15)
    at ServerResponse.send (/mnt/md/tmp/etherpad-anttiviljami/src/node_modules/express/lib/response.js:158:21)
    at ServerResponse.res.send (/mnt/md/tmp/etherpad-anttiviljami/src/node/hooks/express/webaccess.js:105:14)
    at /mnt/md/tmp/etherpad-anttiviljami/src/node/hooks/express/openapi.js:656:17
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
[2020-04-01 13:43:41.607] [ERROR] console - (node:18540) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
[2020-04-01 13:43:41.607] [ERROR] console - (node:18540) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

anttiviljami commented:
Pushed a fix to make error handling for the API nicer. Now there are no errors in the logs anymore.

Also noticed that the API responds with status `200` even when it should clearly return a `400` error. Again, I replicated this weird behaviour to avoid changing tests.

See:
https://github.com/ether/etherpad-lite/commit/247f444966a07240ba5e00cb795f33bd9f61edff#diff-c9c8b1ab998c99480ceda6c011a8420fR681

And:
https://github.com/ether/etherpad-lite/blob/247f444966a07240ba5e00cb795f33bd9f61edff/tests/backend/specs/api/pad.js#L114

anttiviljami commented:
@muxator looks like there's a package-lock.json conflict introduced by adding the http-errors library . I'll let you fix it. 👍

muxator commented:
Github is having issues today.

muxator commented:
Finally conflicts resolved. That was hard.

muxator commented:
With this PR we get rid of `swagger-node-express` (unmantained and with critical vulnerabilities), **do not break compatibility**, and have slightly more test coverage, too.

The ouput of `npm audit` goes from this:

```
added 1003 packages from 1156 contributors and audited 13812 packages in 35.123s
found 7 vulnerabilities (5 low, 2 high)
  run `npm audit fix` to fix them, or `npm audit` for details
```

to this:

```
added 1018 packages from 1164 contributors and audited 13855 packages in 33.844s
found 4 low severity vulnerabilities
  run `npm audit fix` to fix them, or `npm audit` for details
```

**Good work**, @anttiviljami and @JohnMcLear!

anttiviljami commented:
Thanks for merging this PR @muxator @JohnMcLear !

I would really appreciate your honest feedback to help me improve as a developer: https://forms.gle/4GgtN8Se2piaNVjX8
